### PR TITLE
#Sn-15-Create Conversations

### PR DESCRIPTION
**Pull request by ramantechart**

### CODE REVIEW

File: spec/models/conversation_spec.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Conversation, type: :model do
+  subject(:conversation) { FactoryBot.build(:conversation) }
+
+  describe 'validation test' do
```
Comment (mlshvch) [inline spec/models/conversation_spec.rb]: Maybe it's better to rename this describe to `Valid Conversation`

File: spec/models/user_conversation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#  index_user_conversations_on_user_id_and_conversation_id  (user_id,conversation_id) UNIQUE
+#
+require 'rails_helper'
+
+RSpec.describe UserConversation, type: :model do
+  let(:user) { FactoryBot.create(:user) }
+  let(:conversation) { FactoryBot.create(:conversation) }
+  let(:user_conversation) { FactoryBot.create(:user_conversation, user: user, conversation: conversation) }
+
+  describe 'user' do
```
Comment (mlshvch) [inline spec/models/user_conversation_spec.rb]: Please, rename to `#user`

File: spec/models/user_conversation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    context 'with connected user' do
+      it 'must be with correct value' do
+        expect(conversation.user).to eq(user)
+      end
+    end
+  end
+
+  describe 'conversation' do
```
Comment (mlshvch) [inline spec/models/user_conversation_spec.rb]: Please, rename to `#conversation`

**Review by mlshvch — CHANGES_REQUESTED** [review]

File: db/seeds.rb [hunk 4]
```diff
@@ -21,33 +21,38 @@
...
   FactoryBot.create_list(:post_with_pictures_and_likes, GENERATE_POSTS_COUNT, user: user)
 end
 
+USER_SUBSCRIPTIONS_COUNT.times do
+  user1 = users.sample
+  user2 = users.sample
+  loop do
+    user2 = users.sample
+    break if user2 == user1
+  end
```
Comment (ramantechart) [inline db/seeds.rb]:
```
User.all.each_slice(2) do |subscriber, subscription|
  UserSubscription.create(subscriber_id: subscriber.id, subscription_id: subscription.id)
end
```
Probably we can make it way simpler. Let's have just 1 subscriber and 1 subscription

File: spec/factories/conversations.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+FactoryBot.define do
+  factory :conversation do
+    name { Faker::Lorem.word }
+  end
```
Comment (ramantechart) [inline spec/factories/conversations.rb:30]: it misses a trait that creates conversation for multiple users.

File: spec/models/conversation_spec.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+# Table name: conversations
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Conversation, type: :model do
```
Comment (ramantechart) [inline spec/models/conversation_spec.rb:14]: Spec missing case checking users from the conversation.

We need to make sure that `conversation.users` returns only users from the conversation

File: db/seeds.rb (hunk 4 above)
Comment (mlshvch) [reply db/seeds.rb]: I have a proposal:
```
users.each do |user|
  subscribers = users
  subscribers.delete(user)
  rand(0..(NUMBER_OF_USERS-1)).times do
    FactoryBot.create(:user_subscription, subscriber: user, subscription: subscribers.delete(subscribers.sample)) 
    # not sure if it's possible to do the same with create_list
  end
end
```

Comment (mlshvch) [reply db/seeds.rb]: This won't work

File: app/models/user.rb [hunk 7]
```diff
@@ -53,8 +53,11 @@ class User < ApplicationRecord
   has_many :subscriptions, through: :user_subscriptions
   has_many :user_subscribers, class_name: 'UserSubscription', foreign_key: 'subscription_id'
   has_many :subscribers, through: :user_subscribers
+  has_many :user_conversations
+  has_many :users, through: :user_conversations, dependent: :destroy
 
   validates :email, presence: true, uniqueness: true
   validates :status, presence: true, inclusion: { in: %w[active blocked] }
+
```
Comment (mlshvch) [inline app/models/user.rb]: That happened after resolving conflicts, so I think it's not crucial

File: db/migrate/20220124063218_create_messages.rb [hunk 8]
```diff
@@ -6,6 +6,8 @@ def change
     create_table :messages do |t|
       t.text :text, null: false
       t.references :user, null: false, foreign_key: true
+      t.belongs_to :conversation
```
Comment (skripchenkoVeta) [inline db/migrate/20220124063218_create_messages.rb]: Please add the field using another migration

File: db/seeds.rb [hunk 9]
```diff
@@ -20,33 +20,37 @@
...
   FactoryBot.create_list(:post_with_pictures_and_likes, GENERATE_POSTS_COUNT, user: user)
 end
 
-USERS_WITH_MULTIPLE_MESSAGES.times do
-  user = users.delete(users.sample)
-  FactoryBot.create_list(:message, rand(MIN_LIMIT_OF_MESSAGES..MAX_LIMIT_OF_MESSAGES), user: user)
-end
+# rubocop:disable Layout/LineLength
+conversations = FactoryBot.create_list(:conversation, GENERATE_CONVERSATIONS_COUNT, :with_multiple_users, users: users.sample(rand(3..NUMBER_OF_USERS)))
+conversations += FactoryBot.create_list(:conversation, GENERATE_CONVERSATIONS_COUNT, :dialog, users: users)
```
Comment (ramantechart) [inline db/seeds.rb]: this doesn't work as expected. `users` contain more than 2 users

File: db/seeds.rb [hunk 10]
```diff
@@ -20,33 +20,37 @@
...
-UserSubscription.create(subscriber_id: user1, subscription_id: user2)
-
-USERS_WITH_A_SINGLE_MESSAGE.times do
-  user = users.delete(users.sample)
-  FactoryBot.create(:message, user: user)
+conversations.each do |conversation|
+  users = conversation.users
+  users.each do |user|
+    chance = rand
+    if chance.between?(0, 0.4)
```
Comment (ramantechart) [inline db/seeds.rb]:
```
if(chance < 0.4)
  ...
else
  ...
end
```

File: db/seeds.rb [hunk 11]
```diff
@@ -20,33 +20,37 @@
...
+      FactoryBot.create(:message, user: user, conversation: conversation)
+    elsif chance.between?(0.5, 0.9)
+      FactoryBot.create_list(:message, rand(MIN_LIMIT_OF_MESSAGES..MAX_LIMIT_OF_MESSAGES), user: user, conversation: conversation)
```
Comment (ramantechart) [inline db/seeds.rb]: we can go simpler.
Just 
```
.create_list(:message, rand(1..5) ...)
```

**Review by ramantechart — COMMENTED** [review]
Looks good

**Pull request merged**